### PR TITLE
[RL] spmd_types: fix trainer logprob gradient; route generator TP all-reduce via funcol

### PR DESCRIPTION
**Pull request by pianpwk**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #3822

Two spmd_types correctness fixes for the RL stack, to achieve 10-step bitwise.

components/loss.py: compute_logprobs allgathers vocab-parallel logits, this should be to I, not R, so we don't trigger an allreduce in BWD.

experiments/rl/models/vllm_wrapper.py: the vLLM allreduce patching was only applied to DTensor path, and wasn't bitwise to spmd_type's dist.all_reduce call. patch spmd backend by replacing redistribute(P->I) in inference mode (had to be careful here; patching funcol generally would trigger a BWD allreduce, affecting the trainer path, and using P->R semantics).

With both, spmd_types matches the default backend bitwise for the RL trainer after 20 steps for backend in {default, spmd_types}, using async script but no off-policy steps:
```
    python -m torchtitan.experiments.rl.train --module alphabet_sort \
      --config rl_grpo_qwen3_0_6b_flex_batch_invariant \
      --trainer.parallelism.spmd_backend $backend \
      --generator.parallelism.spmd_backend $backend \
      --async_loop.max_offpolicy_steps 0
```

Metrics compared:
```
  Loss / policy-gradient (4) 
  - loss/mean, loss/ratio_mean, loss/ratio_clipped_frac, loss/generator_logprob_nan_frac

  Bitwise diff probes (3) — the framework's own trainer-vs-generator logprob checks
  - bit_wise/logprob_diff/{max,mean}, bit_wise/ratio_tokens_different/mean

  Advantages (5)
  - advantage/{_max,_mean,_min,_std,_sum}

  Reward (7)
  - rollout_reward/{_max,_mean,_min,_std,_sum}, rollout_reward/component/RewardAlphabetSort/mean, rollout_reward/group_zero_std_frac/mean

  Rollout structure / generated sequences (17)
  - rollout/output_tokens/{max,mean,std}, rollout/response_length/{max,mean}, rollout/prompt_length/{max,mean}, rollout/total_length/{max,mean}, rollout/num_turns/{max,mean}, rollout/branches_per_rollout/{max,mean}, rollout/truncation_rate/mean, rollout/{max,min}_policy_version

  Optimizer / training state (4)
  - train/grad_norm/mean, train/lr, train/policy_version, train_batch/policy_age/{mean,max}

  Batch composition (8)
  - train_batch/{num_microbatches,num_rollout_groups,num_training_samples,num_metric_only_groups,padding_frac}, training_sample_builder/{num_training_samples/sum,num_groups_dropped_zero_std/sum}

  Scheduling / buffer state (8) — deterministic in sync mode, also matched
  - rollout_buffer/{active_slots_in_use_peak,available_active_slots,num_groups_finalized,num_groups_inflight,num_groups_waiting}, generator/inflight_requests_at_completion/{max,mean}, generator/num_cached_tokens/{max,mean}
```

Authored with assistance from Claude, an AI coding assistant.

### CODE REVIEW

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -395,8 +395,7 @@ def compute_logprobs(
             logits,
             mesh.get_group("tp"),
             src=spmd.S(-1),
-            dst=spmd.R,
-            backward_options={"op_dtype": logits.dtype},
+            dst=spmd.I,
```
Comment (pianpwk) [inline torchtitan/components/loss.py:398]: logits should be I

**Review by tianyu-l — APPROVED** [review]
